README
NAME
    Test::Reporter::Transport::Socket - Simple socket transport for
    Test::Reporter

VERSION
    version 0.30

SYNOPSIS
        my $report = Test::Reporter->new(
            transport => 'Socket',
            transport_args => [
              host     => 'metabase.relay.example.com',
              port     => 58008,
            ],
        );

        # use space-separated in a CPAN::Reporter config.ini
        transport = Socket host metabase.relay.example.com ...

DESCRIPTION
    This is a Test::Reporter::Transport that sends test report data
    serialised over a TCP socket.

    The idea is to keep dependencies in the tester perl to a bear minimum
    and offload the bulk of submission to a Metabase instance to a relay.

USAGE
    See Test::Reporter and Test::Reporter::Transport for general usage
    information.

  Transport arguments
    Unlike most other Transport classes, this class requires transport
    arguments to be provided as key-value pairs:

        my $report = Test::Reporter->new(
            transport => 'Socket',
            transport_args => [
              host     => 'metabase.relay.example.com',
              port     => 58008,
            ],
        );

    Arguments include:

    "host" (required)
        The name or IP address of a host where we want to send our
        serialised data.

        This may also be an arrayref of the above. A connection will be
        attempted to each item in turn until a successful connection is
        made.

    "port"
        The TCP port on the above "host" to send our serialised data.

METHODS
    These methods are only for internal use by Test::Reporter.

  new
        my $sender = Test::Reporter::Transport::Socket->new( %params );

    The "new" method is the object constructor.

  send
        $sender->send( $report );

    The "send" method transmits the report.

AUTHORS
      David A. Golden (DAGOLDEN)
      Richard Dawe (RICHDAWE)
      Chris Williams (BINGOS)

      This module inlines Config::Perl::V Copyright (C) 2009-2010 H.Merijn Brand

AUTHOR
    Chris Williams <chris@bingosnet.co.uk>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by David A. Golden, Richard Dawe,
    Chris Williams and H.Merijn Brand.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==================================================
Changes from 2009-08-02 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.30 at 2014-08-01 12:30:14 +0000
-----------------------------------------

  Change: d0403d1e1a49240481b746b928df7b850ce18d4d
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2014-08-01 13:30:14 +0000

    Update included Config::Perl::V to CPAN version 0.21 

-----------------------------------------
version 0.28 at 2014-02-19 16:46:43 +0000
-----------------------------------------

  Change: e3805b7982d73e887d991d14e19d327a61a507d9
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2014-02-19 16:46:43 +0000

    Update included Config::Perl::V to CPAN version 0.20 

-----------------------------------------
version 0.26 at 2013-08-01 09:10:18 +0000
-----------------------------------------

  Change: 9eeae1c8384ab245e3a48c2ca43b0c9924471727
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-08-01 10:10:18 +0000

    Remove the Test::Compile usage as this was pulling in Capture::Tiny 

-----------------------------------------
version 0.24 at 2013-07-30 13:19:08 +0000
-----------------------------------------

  Change: 6db2339d9cbcf12a024bf83eff13c942885cdf29
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-07-30 14:19:08 +0000

    Update included Config::Perl::V to CPAN version 0.19 

-----------------------------------------
version 0.22 at 2013-05-24 17:24:23 +0000
-----------------------------------------

  Change: 96fee34be6a2c2b144010d6cca2a0dd46d2d8f2d
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-05-24 18:24:23 +0000

    Update inlined Config::Perl::V to 0.18 

-----------------------------------------
version 0.20 at 2013-02-10 23:04:00 +0000
-----------------------------------------

  Change: 276207efb2d702eb95c7cd2e16201e3d9e91e128
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-02-10 23:04:00 +0000

    Updated included Config::Perl::V to latest version 0.17 

-----------------------------------------
version 0.18 at 2013-01-25 16:57:48 +0000
-----------------------------------------

  Change: 1cf389340e1e9e7b991e389eb95440b20f1ff41b
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-01-25 16:57:48 +0000

    Updated included Config::Perl::V to latest version 

-----------------------------------------
version 0.16 at 2010-10-04 11:59:57 +0000
-----------------------------------------

  Change: f34dcadadeff787bb8b479fecf79db184d5dcc99
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-10-04 12:59:57 +0000

    Tony Cook suggested buffering was the problem behind the errors with
    metabase-relayd, so we check everything has flushed before closing
    the socket now 

-----------------------------------------
version 0.14 at 2010-08-29 17:56:24 +0000
-----------------------------------------

  Change: 65d54a1f1bf50ce36e6b74529ad1b6cf95d0ccce
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-08-29 18:56:24 +0000

    Added diagnostics when we are sending frozen data over the socket to
    the relay 

-----------------------------------------
version 0.12 at 2010-07-06 19:35:14 +0000
-----------------------------------------

  Change: 72c6f4bf9612d2cef9226fe34d245e50428b98b0
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-07-06 20:35:14 +0000

    Switched dist to Dist::Zilla 

  Change: d895b8d6d3463a16fb4a1d8fb53a2444901aca90
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2010-07-06 16:53:17 +0000

    Added arrayref option to 'host' parameter. 

-----------------------------------------
version 0.10 at 2010-03-28 11:07:44 +0000
-----------------------------------------

  Change: d2ce7324ad3376627766e8ec5383a02816e2c1dc
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-03-28 12:07:44 +0000

    This be 0.10 

  Change: 69f5163e1ba4d51d95981ceb385e3f552bc3c663
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-03-28 12:06:58 +0000

    Updated Changes 

  Change: 67b5beb75535280661abbf518b61c1553eab6cd2
  Author: David Golden <dagolden@cpan.org>
  Date : 2010-03-26 13:21:51 +0000

    use Test::Reporter 1.57 style perl version if available 

-----------------------------------------
version 0.08 at 2010-03-23 18:17:38 +0000
-----------------------------------------

  Change: a6b37fea9ab3e6f88fe90544b11bd8dbcf6c6f28
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-03-23 18:17:38 +0000

    Renamed the inline Config::Perl::V to TRTS::Config::Perl::V so that
    PAUSE indexer doesn't squeal like a stabbed pig at us 

-----------------------------------------
version 0.06 at 2010-03-22 21:20:14 +0000
-----------------------------------------

  Change: 8ee597d2ab70d90ed83c780507221ef90b57f0cf
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-03-22 21:20:14 +0000

    Inlined Config::Perl::V 

-----------------------------------------
version 0.04 at 2010-03-17 10:11:03 +0000
-----------------------------------------

  Change: 1065d63d9ffd570c84196e6515bbf8595f0a3ae8
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-03-17 10:11:03 +0000

    Had neglected to add distfile to the data we send. 

-----------------------------------------
version 0.02 at 2010-03-15 23:28:50 +0000
-----------------------------------------

  Change: 3e825954a559f56188f9909336b480f8bd34ccf4
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2010-03-15 23:28:50 +0000

    Moo 

================
End of releases.
================


lib/Test/Reporter/Transport/Socket.pm
package Test::Reporter::Transport::Socket;
$Test::Reporter::Transport::Socket::VERSION = '0.30';
# ABSTRACT: Simple socket transport for Test::Reporter

use strict;
use warnings;
use Carp ();
use IO::Socket::INET;
use Storable qw[nfreeze];
use base qw[Test::Reporter::Transport];

my @required_args = qw/host port/;

sub new {
  my $class = shift;
  Carp::confess __PACKAGE__ . " requires transport args in key/value pairs\n"
    if @_ % 2;
  my %args = @_;
  $args{lc $_} = delete $args{$_} for keys %args;

  for my $k ( @required_args ) {
    Carp::confess __PACKAGE__ . " requires $k argument\n"
      unless exists $args{$k};
  }

  if ( ref $args{host} eq 'ARRAY' and !scalar @{ $args{host} } ) {
    Carp::confess __PACKAGE__ . " requires 'host' argument to have elements if it is an arrayref\n";
  }

  return bless \%args => $class;
}

sub send {
  my ($self, $report) = @_;

  unless ( eval { $report->distfile } ) {
    Carp::confess __PACKAGE__ . ": requires the 'distfile' parameter to be set\n"
      . "Please update your CPAN testing software to a version that provides \n"
      . "this information to Test::Reporter.  Report will not be sent.\n";
  }

  # Open the socket to the given host:port
  # confess on failure.

  my $sock;

  foreach my $host ( ( ref $self->{host} eq 'ARRAY' ? @{ $self->{host} } : $self->{host} ) ) {
    $sock = IO::Socket::INET->new(
      PeerAddr => $host,
      PeerPort => $self->{port},
      Proto    => 'tcp'
    );
    last if $sock;
  }

  unless ( $sock ) {
    Carp::confess __PACKAGE__ . ": could not connect to '$self->{host}' '$!'\n";
  }

  # Get facts about Perl config that Test::Reporter doesn't capture
  # Unfortunately we can't do this from the current perl in case this
  # is a report regenerated from a file and isn't the perl that the report
  # was run on
  my $perlv = $report->{_perl_version}->{_myconfig};
  my $config = TRTS::Config::Perl::V::summary(TRTS::Config::Perl::V::plv2hash($perlv));
  my $perl_version = $report->{_perl_version}{_version} || $config->{version};

  my $data = {
    distfile      => $report->distfile,
    grade         => $report->grade,
    osname        => $config->{osname},
    osversion     => $report->{_perl_version}{_osvers},
    archname      => $report->{_perl_version}{_archname},
    perl_version  => $perl_version,
    textreport    => $report->report
  };

  my $froze;
  eval { $froze = nfreeze( $data ); };

  Carp::confess __PACKAGE__ . ": Could not freeze data '$@'\n"
    unless $froze;

  # Thanks to Tony Cook for this

  while ( length( $froze ) ) {
    my $sent = $sock->send( $froze ) or Carp::confess "Could not send data '$!'\n";
    substr( $froze, 0, $sent, '' );
  }

  close $sock;
  return 1;
}

package TRTS::Config::Perl::V;
$TRTS::Config::Perl::V::VERSION = '0.30';
use strict;
use warnings;

use Config;
use Exporter;
use vars qw(@ISA @EXPORT_OK %EXPORT_TAGS);
@ISA         = ("Exporter");
@EXPORT_OK   = qw( plv2hash summary myconfig signature );
%EXPORT_TAGS = (
    all => [ @EXPORT_OK  ],
    sig => [ "signature" ],
    );

#  Characteristics of this binary (from libperl):
#    Compile-time options: DEBUGGING PERL_DONT_CREATE_GVSV PERL_MALLOC_WRAP
#                          USE_64_BIT_INT USE_LARGE_FILES USE_PERLIO

# The list are as the perl binary has stored it in PL_bincompat_options
#  search for it in
#   perl.c line 1643 S_Internals_V ()
#     perl -ne'(/^S_Internals_V/../^}/)&&s/^\s+"( .*)"/$1/ and print' perl.c
#   perl.h line 4566 PL_bincompat_options
#     perl -ne'(/^\w.*PL_bincompat/../^\w}/)&&s/^\s+"( .*)"/$1/ and print' perl.h
my %BTD = map { $_ => 0 } qw(

    DEBUGGING
    NO_HASH_SEED
    NO_MATHOMS
    NO_TAINT_SUPPORT
    PERL_BOOL_AS_CHAR
    PERL_DISABLE_PMC
    PERL_DONT_CREATE_GVSV
    PERL_EXTERNAL_GLOB
    PERL_HASH_FUNC_DJB2
    PERL_HASH_FUNC_MURMUR3
    PERL_HASH_FUNC_ONE_AT_A_TIME
    PERL_HASH_FUNC_ONE_AT_A_TIME_HARD
    PERL_HASH_FUNC_ONE_AT_A_TIME_OLD
    PERL_HASH_FUNC_SDBM
    PERL_HASH_FUNC_SIPHASH
    PERL_HASH_FUNC_SUPERFAST
    PERL_IS_MINIPERL
    PERL_MALLOC_WRAP
    PERL_MEM_LOG
    PERL_MEM_LOG_ENV
    PERL_MEM_LOG_ENV_FD
    PERL_MEM_LOG_NOIMPL
    PERL_MEM_LOG_STDERR
    PERL_MEM_LOG_TIMESTAMP
    PERL_NEW_COPY_ON_WRITE
    PERL_PERTURB_KEYS_DETERMINISTIC
    PERL_PERTURB_KEYS_DISABLED
    PERL_PERTURB_KEYS_RANDOM
    PERL_PRESERVE_IVUV
    PERL_RELOCATABLE_INCPUSH
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    UNLINK_ALL_VERSIONS
    USE_ATTRIBUTES_FOR_PERLIO
    USE_FAST_STDIO
    USE_HASH_SEED_EXPLICIT
    USE_LOCALE
    USE_LOCALE_CTYPE
    USE_PERL_ATOF
    USE_SITECUSTOMIZE

    DEBUG_LEAKING_SCALARS
    DEBUG_LEAKING_SCALARS_FORK_DUMP
    DECCRTL_SOCKETS
    FAKE_THREADS
    FCRYPT
    HAS_TIMES
    HAVE_INTERP_INTERN
    MULTIPLICITY
    MYMALLOC
    PERL_DEBUG_READONLY_COW
    PERL_DEBUG_READONLY_OPS
    PERL_GLOBAL_STRUCT
    PERL_GLOBAL_STRUCT_PRIVATE
    PERL_IMPLICIT_CONTEXT
    PERL_IMPLICIT_SYS
    PERLIO_LAYERS
    PERL_MAD
    PERL_MICRO
    PERL_NEED_APPCTX
    PERL_NEED_TIMESBASE
    PERL_OLD_COPY_ON_WRITE
    PERL_POISON
    PERL_SAWAMPERSAND
    PERL_TRACK_MEMPOOL
    PERL_USES_PL_PIDSTATUS
    PL_OP_SLAB_ALLOC
    THREADS_HAVE_PIDS
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_IEEE
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE_COLLATE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_LONG_DOUBLE
    USE_PERLIO
    USE_REENTRANT_API
    USE_SFIO
    USE_SOCKS
    VMS_DO_SOCKETS
    VMS_SHORTEN_LONG_SYMBOLS
    VMS_SYMBOL_CASE_AS_IS
    );

# These are all the keys that are
# 1. Always present in %Config - lib/Config.pm #87 tie %Config
# 2. Reported by 'perl -V' (the rest)
my @config_vars = qw(

    api_subversion
    api_version
    api_versionstring
    archlibexp
    dont_use_nlink
    d_readlink
    d_symlink
    exe_ext
    inc_version_list
    ldlibpthname
    patchlevel
    path_sep
    perl_patchlevel
    privlibexp
    scriptdir
    sitearchexp
    sitelibexp
    subversion
    usevendorprefix
    version

    git_commit_id
    git_describe
    git_branch
    git_uncommitted_changes
    git_commit_id_title
    git_snapshot_date

    package revision version_patchlevel_string

    osname osvers archname
    myuname
    config_args
    hint useposix d_sigaction
    useithreads usemultiplicity
    useperlio d_sfio uselargefiles usesocks
    use64bitint use64bitall uselongdouble
    usemymalloc bincompat5005

    cc ccflags
    optimize
    cppflags
    ccversion gccversion gccosandvers
    intsize longsize ptrsize doublesize byteorder
    d_longlong longlongsize d_longdbl longdblsize
    ivtype ivsize nvtype nvsize lseektype lseeksize
    alignbytes prototype

    ld ldflags
    libpth
    libs
    perllibs
    libc so useshrplib libperl
    gnulibc_version

    dlsrc dlext d_dlsymun ccdlflags
    cccdlflags lddlflags
    );

my %empty_build = (
    osname  => "",
    stamp   => 0,
    options => { %BTD },
    patches => [],
    );

sub _make_derived
{
    my $conf = shift;

    for ( [ lseektype   => "Off_t"  ],
    [ myuname   => "uname"  ],
    [ perl_patchlevel => "patch"  ],
    ) {
  my ($official, $derived) = @$_;
  $conf->{config}{$derived}  ||= $conf->{config}{$official};
  $conf->{config}{$official} ||= $conf->{config}{$derived};
  $conf->{derived}{$derived} = delete $conf->{config}{$derived};
  }

    if (exists $conf->{config}{version_patchlevel_string} &&
       !exists $conf->{config}{api_version}) {
  my $vps = $conf->{config}{version_patchlevel_string};
  $vps =~ s{\b revision   \s+ (\S+) }{}x and
      $conf->{config}{revision}        ||= $1;

  $vps =~ s{\b version    \s+ (\S+) }{}x and
      $conf->{config}{api_version}     ||= $1;
  $vps =~ s{\b subversion \s+ (\S+) }{}x and
      $conf->{config}{subversion}      ||= $1;
  $vps =~ s{\b patch      \s+ (\S+) }{}x and
      $conf->{config}{perl_patchlevel} ||= $1;
  }

    ($conf->{config}{version_patchlevel_string} ||= join " ",
  map  { ($_, $conf->{config}{$_} ) }
  grep {      $conf->{config}{$_}   }
  qw( api_version subversion perl_patchlevel )) =~ s/\bperl_//; 

    $conf->{config}{perl_patchlevel}  ||= ""; # 0 is not a valid patchlevel

    if ($conf->{config}{perl_patchlevel} =~ m{^git\w*-([^-]+)}i) {
  $conf->{config}{git_branch}   ||= $1;
  $conf->{config}{git_describe} ||= $conf->{config}{perl_patchlevel};
  }

    $conf;
    } # _make_derived

sub plv2hash
{
    my %config;

    my $pv = join "\n" => @_;

    if ($pv =~ m/^Summary of my\s+(\S+)\s+\(\s*(.*?)\s*\)/m) {
  $config{"package"} = $1;
  my $rev = $2;
  $rev =~ s/^ revision \s+ (\S+) \s*//x and $config{revision} = $1;
  $rev and $config{version_patchlevel_string} = $rev;
  my ($rel) = $config{"package"} =~ m{perl(\d)};
  my ($vers, $subvers) = $rev =~ m{version\s+(\d+)\s+subversion\s+(\d+)};
  defined $vers && defined $subvers && defined $rel and
      $config{version} = "$rel.$vers.$subvers";
  }

    if ($pv =~ m/^\s+(Snapshot of:)\s+(\S+)/) {
  $config{git_commit_id_title} = $1;
  $config{git_commit_id}       = $2;
  }

    if (my %kv = ($pv =~ m/\b(\w+)\s*=\s*('[^']+?'|\S+)/g)) {

  while (my ($k, $v) = each %kv) {
      $k =~ s/\s+$//;
      $v =~ s/,$//;
      $v =~ m/^'(.*)'$/ and $v = $1;
      $v =~ s/^\s+//;
      $v =~ s/\s+$//;
      $config{$k} = $v;
      }
  }

    my $build = { %empty_build };

    $pv =~ m{^\s+Compiled at\s+(.*)}m
  and $build->{stamp}   = $1;
    $pv =~ m{^\s+Locally applied patches:(?:\s+|\n)(.*)}m
  and $build->{patches} = [ split m/\n+/, $1 ];
    $pv =~ m{^\s+Compile-time options:(?:\s+|\n)(.*)}m
  and map { $build->{options}{$_} = 1 } split m/\s+|\n/ => $1;

    $build->{osname} = $config{osname};
    $pv =~ m{^\s+Built under\s+(.*)}m
  and $build->{osname}  = $1;
    $config{osname} ||= $build->{osname};

    return _make_derived ({
  build   => $build,
  environment => {},
  config    => \%config,
  derived   => {},
  inc   => [],
  });
    } # plv2hash

sub summary
{
    my $conf = shift || myconfig ();
    ref $conf eq "HASH" &&
  exists $conf->{config} && exists $conf->{build} or return;

    my %info = map {
  exists $conf->{config}{$_} ? ( $_ => $conf->{config}{$_} ) : () }
  qw( archname osname osvers revision patchlevel subversion version
      cc ccversion gccversion config_args inc_version_list
      d_longdbl d_longlong use64bitall use64bitint useithreads
      uselongdouble usemultiplicity usemymalloc useperlio useshrplib 
      doublesize intsize ivsize nvsize longdblsize longlongsize lseeksize
      );
    $info{$_}++ for grep { $conf->{build}{options}{$_} } keys %{$conf->{build}{options}};

    return \%info;
    } # summary

sub signature
{
    eval { require Digest::MD5 };
    $@ and return "00000000000000000000000000000000";

    my $conf = shift || summary ();
    delete $conf->{config_args};
    return Digest::MD5::md5_hex (join "\xFF" => map {
  "$_=".(defined $conf->{$_} ? $conf->{$_} : "\xFE");
  } sort keys %$conf);
    } # signature

sub myconfig
{
    my $args = shift;
    my %args = ref $args eq "HASH"  ? %$args :
               ref $args eq "ARRAY" ? @$args : ();

    my $build = { %empty_build };

    # 5.14.0 and later provide all the information without shelling out
    my $stamp = eval { Config::compile_date () };
    if (defined $stamp) {
  $stamp =~ s/^Compiled at //;
  $build->{osname}      = $^O;
  $build->{stamp}       = $stamp;
  $build->{patches}     =     [ Config::local_patches () ];
  $build->{options}{$_} = 1 for Config::bincompat_options (),
              Config::non_bincompat_options ();
  }
    else {
  #y $pv = qx[$^X -e"sub Config::myconfig{};" -V];
  my $cnf = plv2hash (qx[$^X -V]);

  $build->{$_} = $cnf->{build}{$_} for qw( osname stamp patches options );
  }

    my @KEYS = keys %ENV;
    my %env  =
  map {      $_ => $ENV{$_} } grep m/^PERL/      => @KEYS;
    $args{env} and
  map { $env{$_} = $ENV{$_} } grep m{$args{env}} => @KEYS;

    my %config = map { $_ => $Config{$_} } @config_vars;

    return _make_derived ({
  build   => $build,
  environment => \%env,
  config    => \%config,
  derived   => {},
  inc   => \@INC,
  });
    } # myconfig

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Reporter::Transport::Socket - Simple socket transport for Test::Reporter

=head1 VERSION

version 0.30

=head1 SYNOPSIS

    my $report = Test::Reporter->new(
        transport => 'Socket',
        transport_args => [
          host     => 'metabase.relay.example.com',
          port     => 58008,
        ],
    );

    # use space-separated in a CPAN::Reporter config.ini
    transport = Socket host metabase.relay.example.com ...

=head1 DESCRIPTION

This is a L<Test::Reporter::Transport> that sends test report data serialised
over a TCP socket.

The idea is to keep dependencies in the tester perl to a bear minimum and offload
the bulk of submission to a Metabase instance to a relay.

=head1 USAGE

See L<Test::Reporter> and L<Test::Reporter::Transport> for general usage
information.

=head2 Transport arguments

Unlike most other Transport classes, this class requires transport arguments
to be provided as key-value pairs:

    my $report = Test::Reporter->new(
        transport => 'Socket',
        transport_args => [
          host     => 'metabase.relay.example.com',
          port     => 58008,
        ],
    );

Arguments include:

=over

=item C<host> (required)

The name or IP address of a host where we want to send our serialised data.

This may also be an arrayref of the above. A connection will be attempted to each
item in turn until a successful connection is made.

=item C<port>

The TCP port on the above C<host> to send our serialised data.

=back

=head1 METHODS

These methods are only for internal use by Test::Reporter.

=head2 new

    my $sender = Test::Reporter::Transport::Socket->new( %params );

The C<new> method is the object constructor.

=head2 send

    $sender->send( $report );

The C<send> method transmits the report.

=head1 AUTHORS

  David A. Golden (DAGOLDEN)
  Richard Dawe (RICHDAWE)
  Chris Williams (BINGOS)

  This module inlines Config::Perl::V Copyright (C) 2009-2010 H.Merijn Brand

=head1 AUTHOR

Chris Williams <chris@bingosnet.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by David A. Golden, Richard Dawe, Chris Williams and H.Merijn Brand.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.006000;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Simple socket transport for Test::Reporter",
  "AUTHOR" => "Chris Williams <chris\@bingosnet.co.uk>",
  "BUILD_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.42",
    "Test::More" => "0.47"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.42"
  },
  "DISTNAME" => "Test-Reporter-Transport-Socket",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Test::Reporter::Transport::Socket",
  "PREREQ_PM" => {
    "IO::Socket::INET" => 0,
    "Storable" => 0,
    "Test::Reporter" => "1.57",
    "Test::Reporter::Transport" => "1.57"
  },
  "VERSION" => "0.30",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => "6.42",
  "IO::Socket::INET" => 0,
  "Storable" => 0,
  "Test::More" => "0.47",
  "Test::Reporter" => "1.57",
  "Test::Reporter::Transport" => "1.57"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



